## Проект по SQL.


#### Задача: 
- Посчитайть количество книг вышедших после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц (исключение брошюр из анализа);
- Определить автора с самой высокой средней оценкой книг (учитываются только книги с 50 и более оценками);
- Посчитайть среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

#### Описание данных:

Таблица `books` - данные о книгах:
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Таблица `authors` - данные об авторах:
- `author_id` — идентификатор автора;
- `author` — имя автора.

Таблица `publishers` - данные об издательствах:
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

Таблица `ratings` - данные о пользовательских оценках книг:
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Таблица `reviews` - данные о пользовательских обзорах на книги:
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

#### Подключение к базе данных:

In [1]:
#!pip install psycopg2-binary

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры

db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

##### Таблица `books`:

In [3]:
query = \
'''

SELECT 
    * 
FROM 
    books
LIMIT 10

'''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

##### Таблица `authors`:

In [4]:
query = \
'''

SELECT * 
FROM authors
LIMIT 10

'''
pd.io.sql.read_sql(query, con = engine)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

##### Таблица `publishers`:

In [5]:
query = \
'''

SELECT * 
FROM publishers
LIMIT 10

'''
pd.io.sql.read_sql(query, con = engine)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

##### Таблица `ratings`:

In [6]:
query = \
'''

SELECT * 
FROM ratings
LIMIT 10

'''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

##### Таблица `reviews`:

In [7]:
query = \
'''

SELECT * 
FROM reviews
LIMIT 10

'''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

##### Подсчёт количества книг, вышедших после 1 января 2000 года:

In [8]:
query = \
'''

SELECT
    COUNT(book_id) AS books_cnt
FROM
    books
WHERE
    publication_date > '2000-01-01'
'''
pd.io.sql.read_sql(query, con = engine)

books_cnt
0        819

##### Рассчёт количества обзоров и средней оценки для каждой книги:

In [9]:
query = \
'''
SELECT 
    books.title, 
    COUNT(distinct reviews.review_id) AS reviews_cnt, 
    AVG(ratings.rating) AS avg_rating
FROM books
LEFT JOIN ratings ON ratings.book_id = books.book_id
LEFT JOIN reviews ON reviews.book_id = books.book_id
GROUP BY 
    books.book_id
ORDER BY 
    COUNT(reviews.review_id) DESC
LIMIT 10
'''
pd.io.sql.read_sql(query, con = engine)

title  reviews_cnt  avg_rating
0                            Twilight (Twilight  #1)            7    3.662500
1                The Hobbit  or There and Back Again            6    4.125000
2                             The Catcher in the Rye            6    3.825581
3  Harry Potter and the Prisoner of Azkaban (Harr...            6    4.414634
4  Harry Potter and the Chamber of Secrets (Harry...            6    4.287500
5               Angels & Demons (Robert Langdon  #1)            5    3.678571
6  Harry Potter and the Order of the Phoenix (Har...            5    4.186667
7  The Lightning Thief (Percy Jackson and the Oly...            6    4.080645
8                                        Animal Farm            5    3.729730
9  The Fellowship of the Ring (The Lord of the Ri...            5    4.391892

##### Издательства, которые выпустило наибольшее число книг толще 50 страниц:

In [10]:
query = \
'''
SELECT
    publisher,
    COUNT(book_id) AS books_cnt
FROM
    publishers
LEFT JOIN books ON  books.publisher_id = publishers.publisher_id 
WHERE 
    num_pages > 50
GROUP BY
    publisher
ORDER BY
    books_cnt DESC
LIMIT 3
'''
pd.io.sql.read_sql(query, con = engine)

publisher  books_cnt
0             Penguin Books         42
1                   Vintage         31
2  Grand Central Publishing         25

##### Авторы с самой высокой средней оценкой книг (оценки > 50):

In [11]:
query = \
'''
SELECT
    author,
    AVG(sub.avg_rating) AS high_avg_rating
FROM
    (SELECT
        author,
        books.book_id,
        AVG(rating) AS avg_rating
    FROM 
        authors
    LEFT JOIN books ON books.author_id = authors.author_id
    LEFT JOIN ratings ON ratings.book_id = books.book_id
    GROUP BY
        author,
        books.book_id
    HAVING
        COUNT(rating) > 50) AS sub
GROUP BY
    author
ORDER BY
    high_avg_rating DESC
'''

pd.io.sql.read_sql(query, con = engine)

author  high_avg_rating
0                          J.K. Rowling/Mary GrandPré         4.283844
1                   Markus Zusak/Cao Xuân Việt Khương         4.264151
2                                      J.R.R. Tolkien         4.258446
3                                   Louisa May Alcott         4.192308
4                                        Rick Riordan         4.080645
5                                     William Golding         3.901408
6                                       J.D. Salinger         3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce         3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...         3.787879
9                                           Dan Brown         3.754540
10                                         Lois Lowry         3.750000
11            George Orwell/Boris Grabnar/Peter Škerl         3.729730
12                                    Stephenie Meyer         3.662500
13                                     John Steinbeck         3.622951

#### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок:

In [12]:
query = \
'''
SELECT
    AVG(sub_review_cnt.review_cnt) AS avg_review_cnt
FROM
    (SELECT
        COUNT(text) AS review_cnt
    FROM 
        reviews
    WHERE username IN
        (SELECT username
        FROM 
            ratings
        GROUP BY
            username
        HAVING
            COUNT(ratings) > 50)
    GROUP BY 
            username
    ) AS sub_review_cnt  
'''

pd.io.sql.read_sql(query, con = engine)

avg_review_cnt
0       24.333333

#### Итог.

- После 1 января 2000 года вышло 819 книг; 
- Для книги `Twilight` вышло 7 обзоров при рейтинге книги	3.7, в то время как для книги `The Hobbit or There and Back Again` при рейтинге 4.1 вышло 6 обзоров;
- Издательство, выпустившее наибольшее число книг толще 50 страниц: `Penguin Books` 42 книги;
- Автор с самой высокой средней оценкой книг: `J.K. Rowling/Mary GrandPré` с рейтингом 4.28;
- Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок: 24.3.